### PR TITLE
Add a receiver for Splunk HEC

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
This adds a Splunk HEC receiver for the OpenTelemetry Collector.
It does not implement the receiver logic itself, and will return errors when used in configuration.

This is a first splintering of the PR #647 

**Testing:** 
Unit tests.

**Documentation:**
README.md and two example files showing the capabilities.

### CODE REVIEW

File: receiver/splunkhecreceiver/config.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"go.opentelemetry.io/collector/config/configtls"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
+)
+
+// Config defines configuration for the SignalFx receiver.
+type Config struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+
+	// TODO: Use one of the configs from core.
```
Comment (bogdandrutu) [inline receiver/splunkhecreceiver/config.go:28]: Can you do that now?

File: receiver/splunkhecreceiver/factory.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+			TypeVal: typeStr,
+			NameVal: typeStr,
+		},
+		Endpoint: defaultEndpoint,
+	}
+}
+
+// extract the port number from string in "address:port" format. If the
+// port number cannot be extracted returns an error.
+func extractPortFromEndpoint(endpoint string) (int, error) {
```
Comment (bogdandrutu) [inline receiver/splunkhecreceiver/factory.go:61]: Do you really need to do this?

**Review by bogdandrutu — APPROVED** [review]

File: receiver/splunkhecreceiver/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+The following settings are optional:
+
+* `access_token_passthrough`: (default = `false`) Whether to preserve incoming
+  access token (`Splunk` header value) as
+  `"com.splunk.hec.access_token"` metric resource label.  Can be used in
+  tandem with identical configuration option for [Splunk HEC
+  exporter](../../exporter/splunkhecexporter/README.md) to preserve datapoint
+  origin.
+* `tls` (no default): This is an optional object used to specify if TLS should be used for
```
Comment (jrcamp) [inline receiver/splunkhecreceiver/README.md:21]: Usually called `tls_settings`.

File: receiver/splunkhecreceiver/factory_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	assert.NotNil(t, cfg, "failed to create default config")
+	assert.NoError(t, configcheck.ValidateConfig(cfg))
+}
+
+type mockMetricsConsumer struct {
+}
+
+var _ consumer.MetricsConsumer = (*mockMetricsConsumer)(nil)
+
+func (m *mockMetricsConsumer) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
```
Comment (jrcamp) [inline receiver/splunkhecreceiver/factory_test.go:42]: Can you use exportertest mocks?

File: receiver/splunkhecreceiver/config.go (hunk 1 above)
Comment (atoulme) [reply receiver/splunkhecreceiver/config.go:28]: looks like I can - changing to adopt it.

File: receiver/splunkhecreceiver/README.md (hunk 3 above)
Comment (atoulme) [reply receiver/splunkhecreceiver/README.md:21]: yes, fixing.

File: receiver/splunkhecreceiver/factory.go (hunk 2 above)
Comment (atoulme) [reply receiver/splunkhecreceiver/factory.go:61]: I lifted this from signalfxreceiver, looks like it's still there. Eventually something that could live in the common HTTPServerSettings struct.

File: receiver/splunkhecreceiver/factory_test.go (hunk 4 above)
Comment (atoulme) [reply receiver/splunkhecreceiver/factory_test.go]: Yes! Fixed.

File: receiver/splunkhecreceiver/factory.go (hunk 2 above)
Comment (bogdandrutu) [reply receiver/splunkhecreceiver/factory.go:67]: I don't understand why extracting the port vs using the endpoint as defined.

Comment (atoulme) [reply receiver/splunkhecreceiver/factory.go:67]: It's just used for validation, apparently. This is something I copied from signalfxreceiver. No strong opinions: if you want it out, I can remove it.

**Comment by bogdandrutu** [comment]
@jrcamp any more comments?

**Review by jrcamp — APPROVED** [review]

**Pull request merged**